Add to box API

khiner opened the pull request:
Similar to @DBraun's [signal API PR](https://github.com/grame-cncm/faust/pull/771).

I'm basically trying to copy the relevant parts of the [`drawSchema` function](https://github.com/grame-cncm/faust/blob/master-dev/compiler/draw/drawschema.hh) for SVGs, as a starting point to be converted over to drawing to other UIs (I'm using ImGui). However, I could use help actually pulling in non-opaque types for `CTree`/`Garbageable`. @DBraun , I see you made `Garbageable` and `CTree` exported in your PR - wondering if you went down a similar path already and got it working?

Note that I can successfully just export/call a `drawBox` method that just delegates to `drawSchema` in the implementation, to generate SVGs.  Also note that I don't need the `gGlobal` instance, since I can do any memoizing/housekeeping myself, but I will need some utilities like `extractName`, `boxpp`, `getDefNameProperty`, etc. I'm assuming those would need to be individually exported & added to `libfaust-box.h`.

Thanks for the help here!

sletz commented:
Thanks for starting this. 

Yes since we decided  to export the `isSigXX` API, then it seems also natural to do that same for the Box API and add the `isBoxXX`functions . But should carefully check what has really to be mode public and what can (or should !?) stay internal.

I'll have a look ASAP (currently doing some deep cleanup on signals in this branch https://github.com/grame-cncm/faust/tree/rework-cast-final4)

khiner commented:
All sounds good, thanks for looking. Of the methods in this PR, `boxComplexity` is the one I'm the least sure about exposing. I'm guessing we'll want to keep that hidden, since it's really only used as a rough internal metric used for splitting up rendering into separate files. I probably won't ultimately use it anyway since I could just render the whole darn nested thing at first, or only show one level at a time, etc. In fact, I think I convinced myself to remove that from this PR :)

DBraun commented:
`Garbageable` and `CTree` are only working in my stuff because I'm doing `#include "tree.h"`, which is bad. One of us should figure out where they should be properly exported in some file: `signals.cpp`, `libcode.cpp`, or other. I'm not sure.

khiner commented:
Thanks, David. After going pretty much the whole way exporting all the various classes/methods needed to replicate all the SVG-drawing functionality, I actually think just importing the needed classes, as you are, probably makes more sense. Implementing a fully-fledged Box/Signal editor/renderer isn't going to be the most common use case to support. I was running into trouble including some compiler files directly but turns out I just needed to include more directories in my CMake. I'm winding back a bunch of `LIBFAUST_API` exports locally (keeping the ones in this PR since they probably still should be exported).

Here's where I'm at so far, likely will bring in more over time:

`CMakeLists.txt` changes:
```
include_directories(
    lib/faust/compiler/errors lib/faust/compiler/tlib lib/faust/compiler/generator
    ...
)
```

Direct includes
```cpp
#include "property.hh"
#include "boxes/boxes.hh"
#include "boxes/ppbox.hh"
#include "utils/names.hh"
#include "description.hh"
```

khiner commented:
Stéphane, when you get to this (no hurry!), I ended up pushing everything I ultimately needed to basically copy over the relevant code from `faust/compiler/draw` to my project ([commit](https://github.com/khiner/flowgrid/commit/55fe9ee06a2de2f64c7a4288508e781cf1e7c87a)) to draw Faust box diagram SVG files.

Overall, I think it strikes a decent balance between:
* using the existing API
* exporting new API code
* including internal Faust headers, and
* copying function implementations

I doubt you'll want to merge this PR as-is, but it's a place to start. Everything's pretty straightforward - the only new API functions that might need explaining are:
```cpp
LIBFAUST_API const char *xtendedName(Tree tree);
LIBFAUST_API unsigned int xtendedArity(Tree tree);
```

`xtendedArity` is there to provide `((xtended )getUserData(tree))->arity()`. Can't just export the `arity` function, since each implementation would need to separately export (I figured that out when I was seeing channel counts in the millions, since it was just interpreting a null implementation as an unsigned int 😆 ).

sletz commented:
So the need is basically "to draw Faust box diagram SVG files." : are we sure exporting all this stuff is the way to go? Or should be design a more high-level and simpler API to cover this need ?

khiner commented:
My goal is to build a graphical editor for Faust. Initially, I'll just be live-rendering a view-only schematic of Faust code to an ImGui window, but ultimately I plan to create a full modular-patching-style visual editor that compiles to the Box/Signal API. I think this is very much in line with what you and the other Faust developers had in mind with the mention of "graphical approaches" in the "Creating a signal language based on this API" section of the  [Signal API docs](https://faustdoc.grame.fr/tutorials/signal-api/), and in the [project ideas](https://github.com/grame-cncm/faustideas/blob/master/README.md#languages-built-on-top-of-the-signal-api) section.

Since the Box API has been opened up since these docs were written, I thought it might be a more natural/direct compilation target than the Signal API for this kind of modular editor application. My thought process in copying over the internals of the SVG-rendering implementation is basically, "where else have I seen schematic renderings of Faust programs, to get started?" I figure I'll need _at least_ as much information from the Box internals as is rendered in the SVGs.

I am very open to feedback/suggestions on the overall approach.

sletz commented:
This seems quite a nice idea yes. Are you on Faust slack?

khiner commented:
I am on the against Slack, yes, and keep an eye on it usually these days. Happy to discuss more there!

khiner commented:
@sletz I've been holding off on this just making sure I won't end up needing anything else, and seeing if you add similar stuff in your refactoring. But I'll just open new PRs for any other needs that come up on my end. Do you think the new API methods this adds are good, and want to merge after I add the missing C-api methods?

khiner commented:
@sletz I've rebased, added C API versions of all exported functions and removed a lot of the methods I originally added. I'm not confident about the C API approach, where every `isBox...` method delegates to the `boxes.hh` version, like many of the C signals API methods.

sletz commented:
@khiner  Some conflicts !

khiner commented:
Ah, looks like from this recent rework of `libcode.cpp` https://github.com/grame-cncm/faust/commit/791a70196131a24cfd8a7f0a36cfcf3de1427f9d. I'll look into re-doing this

khiner commented:
@sletz Updated and fixed my C API methods to use the same pointer-delegating pattern used for the C Signal API

khiner commented on architecture/faust/dsp/libfaust-box-c.h:
```diff
@@ -571,8 +607,62 @@ extern "C"
-    
+    LIBFAUST_API Box CboxAttachAux(Box b1, Box b2);
+
+    LIBFAUST_API bool CisBoxAbstr(Box t, Box& x, Box& y);
```
oops! updated the implementation but not my definitions after resolving the conflicts. changing now...

khiner replied:
Fixed.

khiner commented on architecture/faust/dsp/libfaust-box-c.h:
```diff
@@ -571,8 +607,62 @@ extern "C"
+    LIBFAUST_API bool CisBoxLibrary(Box b, Box& filename);
+    LIBFAUST_API bool CisBoxMerge(Box t, Box& x, Box& y);
+    LIBFAUST_API bool CisBoxMetadata(Box b, Box& exp, Box& mdlist);
+    LIBFAUST_API bool CisBoxModifLocalDef(Box t, Box& body, Box& ldef);
```
No, I don't actually use this one. Deleting now.

khiner replied:
Removed these exports.

khiner commented on architecture/faust/dsp/libfaust-box.h:
```diff
@@ -312,7 +354,7 @@ LIBFAUST_API Box boxSelect2();
  * @return the selected box depending of the selector value at each time t.
  */
-LIBFAUST_API Box boxSelect2(Box selector, Box s1, Box s2);
+LIBFAUST_API Box boxSelect2(Box belector, Box b1, Box b2);
```
good catch!

khiner replied:
fixed.

khiner commented on architecture/faust/dsp/libfaust-box.h:
```diff
@@ -648,4 +690,80 @@ LIBFAUST_API dsp_factory_base* createCPPDSPFactoryFromBoxes(const std::string& n
+LIBFAUST_API bool isBoxNumEntry(Box b, Box& lbl, Box& cur, Box& min, Box& max, Box& step);
+LIBFAUST_API bool isBoxOutputs(Box t, Box& x);
+LIBFAUST_API bool isBoxPar(Box t, Box& x, Box& y);
+LIBFAUST_API bool isBoxPatternMatcher(Box b);
```
I don't use any of these `isBoxPattern...` methods either. Deleting...

khiner replied:
Removed these as well.

khiner commented on compiler/boxes/ppbox.hh:
```diff
@@ -40,17 +40,17 @@
-const char *prim3name(CTree *(*ptr)(CTree *, CTree *, CTree *));
-const char *prim4name(CTree *(*ptr)(CTree *, CTree *, CTree *, CTree *));
-const char *prim5name(CTree *(*ptr)(CTree *, CTree *, CTree *, CTree *, CTree *));
+LIBFAUST_API const char *prim0name(CTree *(*ptr)());
```
I'm using these to get box primitive names for display in the diagram, in exactly the same way as [drawschema](https://github.com/grame-cncm/faust/blob/master-dev/compiler/draw/drawschema.cpp#L424-L436). Open to suggestions to change!

khiner commented on architecture/faust/dsp/libfaust-box.h:
```diff
@@ -83,6 +89,42 @@ extern "C" LIBFAUST_API void createLibContext();
+ *
+ * @return a string representation of a box.
+ */
+LIBFAUST_API const char* tree2str(Box b);
```
Hmm gotcha. This `tree2str` method is already exported with the signal API, but this PR adds the export to the Box API and adds `tree2int` as well, which you pointed out above. Would you suggest passing an error message to be filled, as the `create...` methods do?

khiner replied:
Also, I can just remove these `tree2...` exports from `libfaust-box`, since I can (and do) directly pull them in from `tree.hh`. However, they're still exported with `LIBFAUST_API`, but maybe it's more acceptable if they're not exposed at this level?

khiner commented:
@sletz I think I addressed everything, thanks for the feedback! The only open question is how to deal with the `tree2str`/`tree2int` error handling (or whether to just remove the exports from the box/signal API since I use them directly from `tree.hh` (and maybe @DBraun does too?)

sletz commented:
OK I'll have a second look and @DBraun can comment if this new version fits his need.

sletz commented:
> since I use them directly from `tree.hh` (and maybe @DBraun does too?)

Well you should not.. ((-;  the point of "public" headers is to export considered public stuff ! You should never have to use other headers than the public ones.. Are you using some more "internal" hearder ?

Then I have to find of proper solution for `tree2str/tree2int`

khiner commented:
For sure, I don't feel good about it either :/ I currently pull in `boxes/ppbox.hh` for `boxPrim...` methods/types, and `signals/prim2.hh` for `ffarity` and `ffname` functions. In a separate PR I'd be happy to work on more exports, so these aren't needed.

DBraun commented:
I think I need `tree2str` because I'm using it right now. I don't need `tree2int`. Unfortunately I'm still using `tree.hh` but that doesn't mean that this PR is bad. If it's merged I will continue to be able to do the bad thing of using `tree.hh` but we can figure that out separately.

sletz commented:
Can you guys list all used functions that are not part of the public headers?

khiner commented:
Methods:
* `boxPrim2`
* `boxpp`
* `ffarity`
* `ffname`
* `tree2int`

Types: `prim0`, `prim1`, `prim2`, `prim3`, `prim4`, `prim5`

DBraun commented:
* `tree2str`?


I need tree.hh so that I can use `CTree` and not have this compile error: Pointer to incomplete class type "CTree" is not allowed.

khiner commented:
Ah yes I ran into this too, I knew there was something else important.

sletz commented:
> Pointer to incomplete class type "CTree" is not allowed. 

Where does this happen ?

DBraun commented:
It was a solution to being able to pass Signals/Boxes via pybind11 to the user. I couldn't pass a pointer to a Signal. Instead I had to wrap it somehow.
```c++
struct SigWrapper {
  CTree *ptr;
  SigWrapper(Signal ptr) : ptr{ptr} {}
  operator CTree *() { return ptr; }
};

faust_sig_module.def(
          "sigReal", [](double val) { return SigWrapper(sigReal(val)); },
          arg("val"), returnPolicy)
```

Then in python:
```python
from dawdreamer.faust.signal import *
foo = sigReal(0.5)
```

khiner commented:
And for me, it would take a few minutes to get back to where I can compile my program without any other imports that pull in `tree.hh` to see exactly where it breaks, so I can do that if you'd like in addition to David's case.

sletz commented:
This will help sure. And possibly see what in `tree.hh` is absolutely mandatory to have, so what we separate that part.

khiner commented:
@sletz For my use case, I may actually be able to get away without a concrete `CTree` impl. I pushed up a commit to this PR with an issue I could use some help resolving, related to `ppbox` printing. With this most recent WIP commit, I get the following error
```
Undefined symbols for architecture arm64:
  "boxpp::boxpp(CTree*, int)", referenced from:
      print_tree(CTree*) in DrawBox.cpp.o
ld: symbol(s) not found for architecture arm64
```

Here's the code triggering the error:

```cpp
static string print_tree(Tree tree) {
    stringstream ss;
    ss << boxpp(tree); // <- produces error
    return ss.str();
}
```

I'm clearly not taking the right approach to exposing this `boxpp` constructor. I am guessing the issue is actually defining the class in `libfaust-box`, rather than just declaring an opaque class type. But I'm not sure yet how best to expose this `boxpp(tree)` ctor.

sletz commented:
@khiner  Maybe the best for me would be to directly test on https://github.com/khiner/flowgrid ? Master branch?

khiner commented:
Sounds good, that would probably be easiest - I just pushed up the broken build that produces the error.

DBraun commented on architecture/faust/dsp/libfaust-box.h:
```diff
@@ -648,4 +717,83 @@ LIBFAUST_API dsp_factory_base* createCPPDSPFactoryFromBoxes(const std::string& n
  @}
  */
 
+LIBFAUST_API bool isBoxAbstr(Box t);
```
I think I need all of these functions to change in the same way inside `boxes/boxes.hh`

khiner replied:
Sorry, can you say more about the need/todo here?

khiner commented:
We now have the [khiner/flowgrid:sletz branch](https://github.com/khiner/flowgrid/tree/sletz) pointing to the [khiner/faust:broken](https://github.com/khiner/faust/tree/broken) branch, while this PR's branch (`khiner/faust:master-dev`) is functional with `khiner/flowgrid:main`. (Trying to make things work for @DBraun as well)

sletz commented:
I had a look, but TBH the PR is quite difficult to follow. Let me explain the needs again:

- we want to export what is needed in **public headers** . The point is that a user of libfaust is supposed to **only** use the public headers and link to libfaust. A user should **never** have to include other Faust related files in its code base. We have to follow this in a strict manner. Public API are supposed to stay stable, but everything internal can possibly change, this without breaking external projects

-  we want to be able to use `libfaust-box.h` and `libfaust-signal.h` independently. Here the point is to express that Box and Signal API are consecutive, but separated in the compilation chain. And we want to express that in 2 separated headers

- even if we have 2 separated headers, we need to share some definitions. This is the point of the first common section of `libfaust-box.h` and `libfaust-signal.h` (between `#define LIBFAUSTCOMMON_H` and `#endif`) where you can see the same definition. Everything common should be there

- we already have a `LIBFAUST_API void printSignal(Signal sig, std::ostream& out, bool shared);` that does the job but hide internal stuff

- so we should do the same to print a Box, something like:  `LIBFAUST_API void printBox(Box box, std::ostream& out, bool shared);` and never have to export the internal `boxpp`  stuff.

- can you @khiner rework the PR following those recommendations ? This would greatly help, thanks.

sletz commented:
printBox API added here: https://github.com/grame-cncm/faust/commit/481fadc9ca5cc8e1d69be53e4999315e262c57b3

khiner commented:
@sletz Thanks for your clear explanation and your `printBox` commit.

I've rebased this PR on top of the new `master-dev` to get your `printBox` exports, and added all that was needed for me to build and run successfully with only the following (public) headers imported in my project:

```cpp
#include "faust/dsp/llvm-dsp.h"
#include "faust/dsp/libfaust-signal.h"
#include "faust/dsp/libfaust-signal.h"
```

Please let me know if there are any other concerns with this PR. @DBraun could you please test against this branch (which has been force-pushed again) when you get a chance to see if it still works for you?

DBraun commented:
@khiner Thanks. Everything works for me. The remaining things for me, which we can leave out of this PR, are my issue of including tree.hh and I also want isBoxAdd etc for several bin ops or possibly `LIBFAUST_API bool isBoxBinOp(Box b, Box& x, Box& y, int* i);` where `i` indicates the binary operatr, +, - etc.

sletz commented:
Cleaned a bit and merged in squash mode in https://github.com/grame-cncm/faust/commit/f993b514a0fd69f46cb792b57bc74d959400b95e.

khiner commented:
Thanks Stéphane!
